Add docs for using the family file driver with PyNWB

oruebel opened the pull request:
## Motivation

Fix #1948 

Add docs for using the family file driver with PyNWB


## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured the PR clearly describes the problem and the solution?
- [X] Is your contribution compliant with our coding style? This can be checked running `ruff check . && codespell` from the source directory.
- [X] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [X] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

oruebel committed:
Fix #1948 Add docs for using the family file driver with PyNWB

oruebel committed:
Update changelog

cboulay commented:
Thank you very much for putting this together!

I think any realistic scenario where the user has on disk file size constraints would probably also have in-memory constraints, so it's probably rare that they'd have their entire dataset in memory. I think the provided example would be more practical and more illustrative if the `data` argument was a H5DataIO object with maxshape=(None, nchan).

However, I recognize that this might complicate the message so perhaps it would be enough to just write as much in words. 

> File splitting is most useful with large data and is likely best combined with Iterative Writing. The `data` object in this example can be replaced with DataChunkIterator or H5DataIO. See Iterative Writing documentation for more information.

This glosses over the fact that some of these operations would be done in a different order in an iterative writing scenario but I think that's OK as long as the Iterative Writing documentation is instructive enough.

oruebel committed:
Add iterative write to family example

oruebel commented:
> I think the provided example would be more practical and more illustrative if the `data` argument was a H5DataIO object with maxshape=(None, nchan).

@cboulay I updated the example to include this and then also write data iteratively to that dataset. Can you please take a look and see if that addresses your comment.

bendichter commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,120 @@
+# achieve this by writing different components (e.g., `TimeSeries`) to different files as described above,
+# this option does not allow splitting data from single datasets. An alternative option is to use the
+# `family` driver in `h5py` to automatically split the NWB file into a collection of many HDF5 files.
+# The `family` driver store the file on disk as a series of fixed-length chunks (each in its own file).
```
```suggestion
# The `family` driver stores the file on disk as a series of fixed-length chunks (each in its own file).
```

cboulay commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,120 @@
+# .. note::
+#
+#    The `memb_size` parameter must be set on both write and read. As such, reading the file requires
+#    the user to know the `memb_size` that was used for reading.
```
the user to know the `memb_size` that was used for ~~reading~~ writing.

oruebel replied:
Done

cboulay commented:
There are just a couple small grammar errors noted by me and Ben but otherwise I'd say it's super clear and exactly what I was looking for. Thanks @oruebel!

oruebel committed:
Update plot_linking_data.py

Co-authored-by: Ben Dichter <ben.dichter@gmail.com>

oruebel commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,120 @@
+# .. note::
+#
+#    The `memb_size` parameter must be set on both write and read. As such, reading the file requires
+#    the user to know the `memb_size` that was used for reading.
```
```suggestion
#    the user to know the `memb_size` that was used for writing.
```

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

oruebel commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,120 @@
+# The `family` driver stores the file on disk as a series of fixed-length chunks (each in its own file).
+# In practice, to write very large arrays, we can combine this approach with :ref:`iterative_write` to
+# avoid having to load all data into memory. In the example shown here we use a manual approach to
+# iterative write by using :py:class:`hdmf.backends.hdf5.H5DataIO` to create an empty dataset and
```
```suggestion
# iterative write by using :py:class:`~hdmf.backends.hdf5.h5_utils.H5DataIO` to create an empty dataset and
```

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

oruebel commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,120 @@
+# .. note::
+#
+#    The `memb_size` parameter must be set on both write and read. As such, reading the file requires
+#    the user to know the `memb_size` that was used for writing.
+#
+# .. note::
+#
+#    The DANDI archive may not support NWB files that are split in this fashion.
+
```
```suggestion
#    the user to know the `memb_size` that was used for writing.
#
# .. note::
#
#    The DANDI archive may not support NWB files that are split in this fashion.
#
# .. note::
#
#    Other file drivers, e.g., `split` or `multi` could be used in a similar fashion.
#    However, not all HDF5 drivers are supported by the the high-level API of  
#   `h5py` and as such may require a bit more complex setup via the the 
#    low-level HDF5 API in `h5py`.
#

```

oruebel committed:
Add note about other drivers

oruebel commented:
@stephprince `ruff` seems to fail due to files in `docs/notebooks` that have not been changed in this PR so this may be a separate issue.

Other than the issue with `ruff` this PR is ready for review.

oruebel committed:
Fix indent error

stephprince commented on docs/gallery/advanced_io/plot_iterative_write.py:
```diff
@@ -17,7 +19,7 @@
 #
 # In the typical write process, datasets are created and written as a whole. In contrast,
-# iterative data write refers to the writing of the contents of a dataset in an incremental,
+# iterative data write refers to the writing of the contents of a dataset in an incremental, f
```
```suggestion
# iterative data write refers to the writing of the contents of a dataset in an incremental,
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+
+
+####################
+# Automatically split large data across multiple HDF5 files
```
```suggestion
# Automatically splitting large data across multiple HDF5 files
```

Small change to match the style of the other sections

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+#
+# For extremely large datasets it can be useful to split data across multiple files, e.g., in cases where
+# the file stystem does not allow for large files. While we can
+# achieve this by writing different components (e.g., `TimeSeries`) to different files as described above,
```
```suggestion
# achieve this by writing different components (e.g., :py:meth:`~pynwb.base.TimeSeries`) to different files as described above,
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+# achieve this by writing different components (e.g., `TimeSeries`) to different files as described above,
+# this option does not allow splitting data from single datasets. An alternative option is to use the
+# `family` driver in `h5py` to automatically split the NWB file into a collection of many HDF5 files.
+# The `family` driver stores the file on disk as a series of fixed-length chunks (each in its own file).
```
```suggestion
# ``family`` driver in ``h5py`` to automatically split the NWB file into a collection of many HDF5 files.
# The ``family`` driver stores the file on disk as a series of fixed-length chunks (each in its own file).
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+# then filling in the data afterward.
+
+####################
+# Step 1: Create the `NWBFile` as usual
```
```suggestion
# Step 1: Create the NWBFile as usual
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+nwbfile.add_acquisition(time_series)
+
+####################
+# Step 2: Open the new file with the `family` driver and write
```
```suggestion
# Step 2: Open the new file with the family driver and write
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+####################
+# Step 2: Open the new file with the `family` driver and write
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+# Here we need to open the file with `h5py` first to set up the driver, and then we can use
+# that file with :py:class:`pynwb.NWBHDF5IO`. This is required, because :py:class:`pynwb.NWBHDF5IO`
+# currently does not support passing the `memb_size` option required by the `family` driver.
```
```suggestion
# Here we need to open the file with ``h5py`` first to set up the driver, and then we can use
# that file with :py:class:`pynwb.NWBHDF5IO`. This is required because :py:class:`pynwb.NWBHDF5IO`
# currently does not support passing the ``memb_size`` option required by the ``family`` driver.
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+# .. note::
+#
+#    Alternatively, we could have also used the :ref:`iterative_write` features to write the data
+#    iteratively directly as part of the `io.write` call instead of manually afterward.
```
```suggestion
#    iteratively directly as part of the ``io.write`` call instead of manually afterward.
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+#    iteratively directly as part of the `io.write` call instead of manually afterward.
+
+####################
+# Step 3: Read a file written with the `family` driver
```
```suggestion
# Step 3: Read a file written with the family driver
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+####################
+# .. note::
+#
+#    The filename you provide when using the `family` driver must contain a printf-style integer format code
```
```suggestion
#    The filename you provide when using the ``family`` driver must contain a printf-style integer format code
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+# .. note::
+#
+#    The `memb_size` parameter must be set on both write and read. As such, reading the file requires
+#    the user to know the `memb_size` that was used for writing.
```
```suggestion
#    The ``memb_size`` parameter must be set on both write and read. As such, reading the file requires
#    the user to know the ``memb_size`` that was used for writing.
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+#    Other file drivers, e.g., `split` or `multi` could be used in a similar fashion.
+#    However, not all HDF5 drivers are supported by the the high-level API of
+#    `h5py` and as such may require a bit more complex setup via the the
+#    low-level HDF5 API in `h5py`.
```
```suggestion
#    Other file drivers, e.g., ``split`` or ``multi`` could be used in a similar fashion.
#    However, not all HDF5 drivers are supported by the the high-level API of
#    ``h5py`` and as such may require a bit more complex setup via the the
#    low-level HDF5 API in ``h5py``.
```

stephprince commented on docs/gallery/advanced_io/plot_linking_data.py:
```diff
@@ -259,5 +259,128 @@
+#    The `memb_size` parameter must be set on both write and read. As such, reading the file requires
+#    the user to know the `memb_size` that was used for writing.
+#
+# .. note::
```
```suggestion
# .. warning::
```

Could consider updating this to a warning to make it stand out more, since the impact might not be noticed until much later after the data is written.

stephprince reviewed (commented):
Looks good and read clearly! I added some small formatting suggestions, mainly some code references that were italicized that I think were meant to be code blocks. If the italics were intentional you can just ignore those suggestions. 

It looks like some of the ruff linter settings were deprecated - I will address those in a separate PR and we can ignore it here for now.

oruebel committed:
Update docs/gallery/advanced_io/plot_iterative_write.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel committed:
Update docs/gallery/advanced_io/plot_linking_data.py

Co-authored-by: Steph Prince <40640337+stephprince@users.noreply.github.com>

oruebel commented:
> I added some small formatting suggestions, mainly some code references that were italicized that I think were meant to be code blocks. If the italics were intentional you can just ignore those suggestions.

Thanks for the suggestions. All those edits looked good and I committed them to this PR.

oruebel commented:
> It looks like some of the ruff linter settings were deprecated - I will address those in a separate PR and we can ignore it here for now.

Sounds good. I think this PR is then ready to be merged.
